cpp/src/arrow/io/interfaces.h

```diff
@@ -43,9 +45,9 @@ class FileSystemClient {
   virtual ~FileSystemClient() {}
 };
 
-class FileInterface {
+class ARROW_EXPORT FileInterface {
  public:
-  virtual ~FileInterface() {}
+  virtual ~FileInterface() = 0;
```

Comment:
I did this on purpose — I had to do some wrangling with the class hierarchy when I added default virtual implementations (which can be overridden, like they are for HDFS) of `ReadAt`